### PR TITLE
cli,server: support non-TLS localhost-only HTTP

### LINKED ISSUE #46420: ui,server: enable pain-free test-on-own-laptop user journey
Today it's hard to set up a secure cluster on a local machine for testing.

It is hard because a localhost-only server would present a localhost-only TLS cert, and [some OS/browser combinations have become very agressive at preventing dubious TLS certs](https://github.com/cockroachdb/docs/issues/6946) for good reasons.

In order to smoothen out the user journey for "testing on your own laptop", this issue proposes to introduce a new command-line flag, for example `--unencrypted-localhost-http` (but also see discussion below)

which, to summarize, enables the user to point their browser to `http://localhost:8080` (note `http://` and not `https://`) and get UI access without cert-related headaches.

This change is a sub-set of the more general approach described by #44842; the subset would be appropriate to back-port to 19.2.

### Discussion

We can approach this either:

1) by making `--unencrypted-localhost-http` **imply** `--http-addr=localhost` and disabling TLS for the http connection.

   In that approach:

   - pros: there's a single URL for all the HTTP endpoints (UI, debug, health etc)
   - cons: *all* the HTTP endpoints (incl debug, health etc) will have no TLS, so *scripts may need to be customized*
   - cons: it's not possible to access the UI from other machines

2) by making `--unencrypted-localhost-http` cause a **second listener** on `localhost`, *in addition* to the main TLS-enabled listener configured by `--http-addr`, using (presumably) a different TCP port.

   In that approach:

   - pros: the debug/health/status etc endpoints continue to use TLS as previously, which preserves compatibility with automation
   - cons: there's a separate URL to point the browser to, separate from that used by automation tools using debug/health etc). The UI URL would still be printed as `webui: http://localhost:xxxx` upon `cockroach start`. The "main" httpS URL would then be printed alongside as `http API: https://something:yyyy` alongside `webui:`.
   - cons: still not possible to access the UI from other machines

3) do not introduce a new flag, instead extend `--http-addr` to become a comma-delimited list: `--http-addr=localhost:8080,localhost:8081:notls` (**notice the localhost:8080:notls** at the end).

   In the first implementation, we'd support a maximum of 2 listeners (= either 1 or 2).
 
   In that approach:

   - pros: it's possible to have a single URL for both UI and APIs
   - pros: it's possible to have both TLS on non-localhost, and TLS on localhost, or both on localhost - with a suitable config, we get *both* browser happiness  and script happiness
   - pros: it's one of the 3 steps needed to complete #44842 so it moves our 20.2 roadmap further
   - cons: more complex implementation than the above
   - cons: (possibly) we need to explain why `--http-addr` supports this feature and not the other listen flags.

### PR DESCRIPTION
**Pull request by knz**
Fixes #46420. 

tldr: this patch makes it possible to disable TLS for the HTTP
endpoint and thus facilitate UI access for testing, when proper cert
setup in the web browser is complicated.

Use with: `cockroach start` (or `start-single-node`)
and flag `--unencrypted-localhost-http`. The resulting HTTP URL
uses `http://` and thus no TLS.

**Motivation**

We distinguish two user journeys to discover CockroachDB and "try it
out": setting up a node in the Cloud, and setting up a node on a
personal computer.

The first case requires good TLS security, but then setting up TLS
certificates for a cloud host is relatively straightforward: the Cloud
provider's own cert management interface, Let's Encrypt or manual
configuration using the cloud host's stable hostname are reliable and
well-documented methods.

The second case (own computer) is complicated.

Personal computers typically do not have stable addresses and setting
up a trusted TLS cert for their network address is thus difficult.
Configuring a TLS cert for `localhost` is theoretically possible but
troublesome for the web UI. Indeed, certain OS/browser combinations
have become very agressive at preventing dubious TLS certs over
time (e.g. Chrome on macOS Catalina)—for good reasons from a general
Internet security perspective, but in a way impractical to
CockroachDB.

On the other hand, running a server on the local machine does not
usually come with the requirement that the service is available over
the network for other users. When there is no actual network
communication, transport-level security (i.e. encryption and server
authentication) is not useful.

We're thus interested to acknowledge this last point and this patch
provides this, a feature to *disable TLS for HTTP connections and
restrict the HTTP listener to the loopback interface*, while keeping
the other security mechanisms active (i.e. not force `--insecure`).

**Detail of the changes**

- a new boolean field on `base.Config`: `DisableTLSForHTTP`
- when set, disables loading the UI TLS configuration
  upon `(*Server) Start()`, and thus disables the TLS
  wrapper for the HTTP listener.
- new CLI flag `--unencrypted-localhost-http` forces
  both `base.HTTPAddr` to use `localhost` and
  sets `DisableTLSForHTTP`.
- the TLS-less URL is also advertised in the `NodeInfo`'s
  `AdminURL` field.

Release justification: fixes for high-priority or high-severity bugs in existing functionality

Release note (ui change): a server started with
`--unencrypted-localhost-http` is now reachable with a `http://` URL
in the web browser and does not require certificate or CA setup for
HTTP accesses.

Release note (security update): the new command-line argument
`--unencrypted-localhost-http`, when specified on either `cockroach
start` or `cockroach start-single-node`, forces the HTTP listener to
bind to `localhost` addresses only and disables the TLS protocol. This
effectively disables transport-level security and server
authentication for the HTTP interface.

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-:-M3CCNZl8VEtA-wlFV-l:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jseldess and @knz)

___
*[pkg/cli/flags.go, line 786 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-M3C9ol76Sel5LkwXYfQ:-M3C9ol76Sel5LkwXYfR:bgk51hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7db14bf1f164772ea4cecab3dffb92d5fc6d7c4/pkg/cli/flags.go#L786)):*
> ```Go
> 
> 	// Fill in the defaults for --http-addr.
> 	if serverHTTPAddr == "" {
> ```

Would it be helpful to complain if the user specified `--http-addr`  to something other than localhost and then also specified `--unencrypted-localhost-http`

**Review by jseldess — APPROVED** [review]
:lgtm: thought I haven't done any testing. Would you like me to? I think there are potentially confusing combinations people might run into. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-:-M3CM29d6pGJb7q8Ty4X:bogwxrs)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/cliflags/flags.go, line 345 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-M3CK4DT5BiT3TNk3JXt:-M3CK4DU1OdeMk085WdZ:b-29j3tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7db14bf1f164772ea4cecab3dffb92d5fc6d7c4/pkg/cli/cliflags/flags.go#L345)):*
> ```Go
> 
> </PRE>
> When starting a singe-node cluster, leave this flag out. This will
> ```

Totally unrelated, but should we remove this note now that `start` without `--join` is deprecated?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46471)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner and @jseldess)

___
*[pkg/cli/flags.go, line 786 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-M3C9ol76Sel5LkwXYfQ:-M3CZbyq6tl46OO1EBbO:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7db14bf1f164772ea4cecab3dffb92d5fc6d7c4/pkg/cli/flags.go#L786)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Would it be helpful to complain if the user specified `--http-addr`  to something other than localhost and then also specified `--unencrypted-localhost-http`
</blockquote></details>

Good idea. Done.
___
*[pkg/cli/cliflags/flags.go, line 345 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46471#-M3CK4DT5BiT3TNk3JXt:-M3COK1i5gn5GiSfzhVP:b-70yn68) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7db14bf1f164772ea4cecab3dffb92d5fc6d7c4/pkg/cli/cliflags/flags.go#L345)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Totally unrelated, but should we remove this note now that `start` without `--join` is deprecated?
</blockquote></details>

That PR unfortunately did not make the cut for 20.1 so we're pushing that change one more release.

**Comment by knz** [comment]
Thanks!

bors r=ajwerner,jseldess

**Comment by knz** [comment]
@isaactwong @piyush-singh @jseldess  I have received a request via side channels to also backport this to 19.2. Do you confirm?
